### PR TITLE
Split StorageSlot into TransientSlot

### PR DESCRIPTION
**Pull request by ernestognw**
To avoid disrupting StorageSlot users, we're splitting the transient storage part to another library that's compiled through Solidity ^0.8.24

#### PR Checklist

- [ ] Tests
- [ ] Documentation
- [ ] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: scripts/generate/run.js [hunk 1]
```diff
@@ -39,9 +39,11 @@ for (const [file, template] of Object.entries({
   'utils/structs/EnumerableMap.sol': './templates/EnumerableMap.js',
   'utils/SlotDerivation.sol': './templates/SlotDerivation.js',
   'utils/StorageSlot.sol': './templates/StorageSlot.js',
+  'utils/StorageSlotTransient.sol': './templates/StorageSlotTransient.js',
```
Comment (Amxx) [inline scripts/generate/run.js:42]: I'm think "Transient" and "Storage" should go together. Also this is not a "transeient-variant" of the StorageSlot library (because it uses different type, different functions, ...)

So I'd be in vafor of changing the name. Proposal include:
- [ ] `Transient.sol`
- [ ] `TransientSlot.sol`
- [ ] `TransientStorageSlot.sol`
- [ ] Something else ?

File: contracts/utils/README.adoc [hunk 2]
```diff
@@ -34,7 +34,8 @@ Miscellaneous contracts and libraries containing utility functions you can use t
  * {Strings}: Common operations for strings formatting.
  * {ShortString}: Library to encode (and decode) short strings into (or from) a single bytes32 slot for optimizing costs. Short strings are limited to 31 characters.
  * {SlotDerivation}: Methods for deriving storage slot from ERC-7201 namespaces as well as from constructions such as mapping and arrays.
- * {StorageSlot}: Methods for accessing specific storage slots formatted as common primitive types. Also include primitives for reading from and writing to transient storage (only value types are currently supported).
+ * {StorageSlot}: Methods for accessing specific storage slots formatted as common primitive types. 
+ * {TransientSlot}: primitives for reading from and writing to transient storage (only value types are currently supported).
```
Comment (Amxx) [inline contracts/utils/README.adoc:38]:
```suggestion
 * {TransientSlot}: Primitives for reading from and writing to transient storage (only value types are currently supported).
```

File: lib/forge-std
Comment (Amxx) [inline lib/forge-std:1]: is that change needed here? If not, I'd do it in a separate PR.

**Review by Amxx — DISMISSED** [review]
LGTM

I'd like @frangio opinion, particularly on the UDVT defined and used in TransientSlot.sol

**Comment by Amxx** [comment]
For the record, here is the rational for this change:

- In the upgrade plugins, we compile ERC-1967 targetting paris. The logic is that we don't know which chain is going to be targeted, and weither it supports push0 or not. So we ship "old versions" of the proxy
- This proxy code relies on the StorageSlot library (for reading/writting the implementation slot)
- The compiler will refuse to compile that is we set the evm target to paris and if the library contains references to tstore/tload (even if they are not actually used of the contract being compiled)

Other users may have similar situations where they use `StorageSlot.sol` with 0.8.20~0.8.23 ... or targeting evm version that do not include transient storage. They should be able to continue using `StorageSlot.sol`. 

Overall, no contract should import any reference to `tstore/tload` unless they actually plan on using transient storage.

**Comment by frangio** [comment]
> I'd like @frangio opinion, particularly on the UDVT defined and used in TransientSlot.sol

The type is basically the same, it's just been moved, right?

I've expressed my opinion before that the "transient slot" type should be different from the "(non-transient) storage slot" type. By separating the libraries this PR is closer to that goal, so I think that's good.

**Review by arr00 — APPROVED** [review]

**Review by arr00 — APPROVED** [review]

**Comment by ernestognw** [comment]
For the record, we agreed with the naming and parameters during a call. The last update only fixed the lining rules.
Thanks for approving @arr00!
